Vulkan Multiview support for pixel shader debugging

Zorro666 opened the pull request:
## Description
- Add slice information to the Pixel History Window Title if the texture arraysize > 0
- Implement Issue #3097 
  - Output `ViewIndex` into the pixel shader hits array
  - Select the pixel shader input which matches the `viewIndex` to the passed in `view` parameter
  - Ignore the view parameter if set to ~0U (`rd.ReplayController.NoPreference`) 
  - Ignore the view parameter if any subpass has an empty multiviews

## Testing
- Tested using Sascha Williams `multiview` sample
- New Automated Test `VK_Multi_View`
  - Vertex shader output Red for view 0, Green for view 1, fragment shader is pass thru and does not use viewIndex
  - Fragment shader output Red for view 0, Green for view 1, vertex shader does not use viewIndex
  - Geometry shader output Red for view 0, Green for view 1, fragment shader is pass thru and does not use viewIndex
  - Vertex shader output fixed colour to pass thru fragment, neither shader uses viewIndex 
  - Not implemented (because Mesh Shader does not currently support multiview): Mesh shader output Red for view 0, Green for view 1, fragment shader does not use viewIndex 
  - Python checks vertex and pixel shader debug output against replay rendering output
- Ran D3D11_, D3D12_, VK_ automated tests locally to check the python compiles with the extra required parameter for `DebugPixel`

Zorro666 committed:
Add slice and instid parameters to DebugPixel()

instid is not currently supported but is there to reduce the amount of API change required in the future

Zorro666 committed:
Add slice information to the Pixel History Window Title

Zorro666 committed:
Vulkan DebugPixel support for multiview debugging

Output "ViewIndex" into the pixel shader hits array
Select the pixel shader input which matches the viewIndex to the passed in view
Ignore the view parameter if set to ~0U (rd.ReplayController.NoPreference) 
Ignore the view parameter if any subpass has an empty multiviews

Zorro666 committed:
Vulkan Multi-view test

Vertex shader output Red for view 0, Green for view 1, fragment shader does not use viewIndex.
Fragment shader output Red for view 0, Green for view 1, vertex shader does not use viewIndex.
Geometry shader output Red for view 0, Green for view 1, vertex and fragment shader do not use viewIndex.

Python checks vertex and pixel shader debug output against replay rendering output

baldurk commented on qrenderdoc/Windows/PixelHistoryView.cpp:
```diff
@@ -746,7 +748,7 @@ void PixelHistoryView::startDebug(EventTag tag)
   m_Ctx.Replay().AsyncInvoke([this, &trace, &done, tag](IReplayController *r) {
     trace = r->DebugPixel((uint32_t)m_Pixel.x(), (uint32_t)m_Pixel.y(),
-                          m_Display.subresource.sample, tag.primitive);
+                          m_Display.subresource.sample, tag.primitive, m_Display.subresource.slice);
```
Render target array slices and multiview view indices are unfortunately not interchangeable like this.

Multiview will broadcast to corresponding indices within the attached image view, but the image view may not have its first array slice at slice 0. That will be a common case, but not the only case. If you have an image view bound which has slice 3 as the first slice and 4-way multiview then view index 0 is slice 3, view index 1 is slice 4, etc.

baldurk commented on qrenderdoc/Windows/TextureViewer.cpp:
```diff
@@ -4098,7 +4098,8 @@ void TextureViewer::on_debugPixelContext_clicked()
   m_Ctx.Replay().AsyncInvoke([this, &trace, &done, x, y](IReplayController *r) {
-    trace = r->DebugPixel((uint32_t)x, (uint32_t)y, m_TexDisplay.subresource.sample, ~0U);
+    trace = r->DebugPixel((uint32_t)x, (uint32_t)y, m_TexDisplay.subresource.sample, ~0U,
+                          m_TexDisplay.subresource.slice);
```
This also interchanges slice and view invalidly.

baldurk commented on renderdoc/api/replay/renderdoc_replay.h:
```diff
@@ -970,12 +970,13 @@ bucket when the pixel values are divided between ``minval`` and ``maxval``.
 :param int sample: The multi-sampled sample. Ignored if non-multisampled texture.
 :param int primitive: Debug the pixel from this primitive if there's ambiguity. If set to
   :data:`NoPreference` then a random fragment writing to the given co-ordinate is debugged.
+:param int view: Debug the pixel writing to this view, ignored if set to :data:`NoPreference`.
```
It would be good to be explicit here about this referring only to multiview views (since viewports are a different concept and there's room for ambiguity here). Also mentioning that it's ignored if there is no multiview in use.

baldurk commented on renderdoc/driver/vulkan/vk_shaderdebug.cpp:
```diff
@@ -4295,6 +4369,30 @@ ShaderDebugTrace *VulkanReplay::DebugPixel(uint32_t eventId, uint32_t x, uint32_
+  {
+    ResourceId rp = state.GetRenderPass();
+    const VulkanCreationInfo::RenderPass &rpInfo =
+        m_pDriver->GetDebugManager()->GetRenderPassInfo(rp);
```
This is not valid if dynamic rendering is in use and there is no renderpass. The viewIndex mask will need to be obtained from the dynamic rendering info.

baldurk commented on renderdoc/driver/vulkan/vk_shaderdebug.cpp:
```diff
@@ -4245,6 +4318,7 @@ ShaderDebugTrace *VulkanReplay::DebugPixel(uint32_t eventId, uint32_t x, uint32_
   builtins[ShaderBuiltin::DrawIndex] = ShaderVariable(rdcstr(), action->drawIndex, 0U, 0U, 0U);
   builtins[ShaderBuiltin::Position] =
       ShaderVariable(rdcstr(), float(x) + 0.5f, float(y) + 0.5f, 0.0f, 0.0f);
+  builtins[ShaderBuiltin::ViewportIndex] = ShaderVariable(rdcstr(), view, 0U, 0U, 0U);
```
This is my prior mistake, not a problem in this PR, but I realise now that `ViewIndex` shouldn't be mapped to the same builtin as `ViewportIndex`. They're two different builtins with two different meanings.

There should be a new `ShaderBuiltin` enum value called I think to be easier to identify `ShaderBuiltin::MultiViewIndex` which should be used in the places you're using it here.

Also - I think this builtin should only be set if we have a view we're looking for, not if its `~0U`. In most cases it should get overwritten by the read value anyway below when we fetch PS inputs, but we should be safe.

baldurk commented on renderdoc/driver/vulkan/vk_shaderdebug.cpp:
```diff
@@ -4714,6 +4812,13 @@ ShaderDebugTrace *VulkanReplay::DebugPixel(uint32_t eventId, uint32_t x, uint32_
+    {
+      if(hit->view != view)
+        continue;
+    }
```
I think this should use the same logic as below which prioritises getting a hit over the selected hit. E.g. for primitives we'll debug a different primitive if there are no hits from the selected one.

baldurk reviewed (changes requested):
Overall this looks good but I've noted a few problems - one of which is mine which this PR highlights.

One thing that occurs to me following up from one of the notes - for the public API is it better to have the view as the parameter or the slice? They are related and 1:1 but not the same - slice is a more intuitive thing to want to select by, but this improvement implements a common special-case for finding the slice via the view index. In theory though someone could want to do this for regular layered rendering which then wouldn't need a change in the API.

Zorro666 committed:
Add ShaderBuiltin::MultiViewIndex

Mapped to:
SPIR-V ViewIndex (which was previously mapped to ShaderBuiltin::ViewportIndex)
GL gl_ViewIndex

Zorro666 committed:
Fix off by one error in the slice range tooltip

Zorro666 commented:
Closing this PR - the work has grown and will be broken down into multiple (at least two) PRs
